server/blob/blob-api/src/test/java/org/apache/james/blob/api/BucketBlobStoreContract.java

```diff
@@ -0,0 +1,154 @@
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.blob.api;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+
+import org.junit.jupiter.api.Test;
+
+public interface BucketBlobStoreContract extends MetricableBlobStoreContract {
+    BucketName CUSTOM = BucketName.of("custom");
+
+    @Test
+    default void createBucketShouldThrowWhenNullBucketName() {
+        assertThatThrownBy(() -> testee().createBucket(null).block())
+            .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    default void createBucketShouldThrowWhenCreatingExistingBucket() {
+        testee().createBucket(CUSTOM).block();
+        assertThatThrownBy(() -> testee().createBucket(CUSTOM).block())
+            .isInstanceOf(ObjectStoreException.class);
+    }
+
+    @Test
+    default void saveBytesShouldThrowWhenNullBucketName() {
+        assertThatThrownBy(() -> testee().save(null, SHORT_BYTEARRAY).block())
+            .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    default void saveStringShouldThrowWhenNullBucketName() {
+        assertThatThrownBy(() -> testee().save(null, SHORT_STRING).block())
+            .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    default void saveInputStreamShouldThrowWhenNullBucketName() {
+        assertThatThrownBy(() -> testee().save(null, new ByteArrayInputStream(SHORT_BYTEARRAY)).block())
+            .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    default void saveBytesShouldThrowWhenBucketDoesNotExist() {
+        assertThatThrownBy(() -> testee().save(CUSTOM, SHORT_BYTEARRAY).block())
+            .isInstanceOf(ObjectStoreException.class);
+    }
+
+    @Test
+    default void saveStringShouldThrowWhenBucketDoesNotExist() {
+        assertThatThrownBy(() -> testee().save(CUSTOM, SHORT_STRING).block())
+            .isInstanceOf(ObjectStoreException.class);
+    }
+
+    @Test
+    default void saveInputStreamShouldThrowWhenBucketDoesNotExist() {
+        assertThatThrownBy(() -> testee().save(CUSTOM, new ByteArrayInputStream(SHORT_BYTEARRAY)).block())
+            .isInstanceOf(ObjectStoreException.class);
+    }
+
+    @Test
+    default void saveShouldReturnBlobIdOfBytesInCreatedBucket() {
+        testee().createBucket(CUSTOM).block();
+        BlobId blobId = testee().save(CUSTOM, SHORT_BYTEARRAY).block();
+
+        assertThat(blobId).isEqualTo(blobIdFactory().from("31f7a65e315586ac198bd798b6629ce4903d0899476d5741a9f32e2e521b6a66"));
```

Comment:
I think this assert would be better, you assert a string equals but with me it doesn't say much, because IDK how was that plain string created:
```java
assertThat(testee().read(CUSTOM, blobId).block())
       .hasSameContentAs(new ByteArrayImputStream(SHORT_BYTEARRAY));
```

BTW I think this kind of test is already existed in BlobStoreContract (saving a blob in a bucket then read it then assert)

Reply:
The string is taken from other contracts, no idea how it is made as well. 

Regarding the tests, in `BlobStoreContract` it's just storing something in the default bucket that has been created along with the BlobStore. This one is checking that we can still save and read a blob in a custom bucket that we manually create before. If I don't test that I don't know what I'm supposed to test with those buckets then?